blockchain/blockchain.go

```diff
@@ -112,7 +112,7 @@ type (
 )
 
 // ProductivityByEpoch returns the map of the number of blocks produced per delegate in an epoch
-func ProductivityByEpoch(ctx context.Context, bc Blockchain, epochNum uint64) (uint64, map[string]uint64, error) {
+func ProductivityByEpoch(ctx context.Context, bc Blockchain, sr protocol.StateReader, epochNum uint64) (uint64, map[string]uint64, error) {
```

Comment:
to my understanding, we shouldn't modify this function

Reply:
In the future after merge with #2016, this function will be only used for tip info. So I pass sr to be sf at the chainservice. Could you explain why I shouldn't modify this func?

Reply:
a generic productivityByEpoch API should only have either sr or epochNum as parameter, but not both. If this is an API provided by blockchain, we may need to skip sr as parameter. Correspondingly we don't need to know who are the delegates by epoch, but just return a static of map[string]int, where string is the producer address in string, int is the number of blocks the producer produced in the epoch. For the delegates who produced 0 blocks, they won't be in the return map.

Reply:
Oh I got it! but if we do that, we have to call Delegates() func in reward protocol and poll protocol to calculate unproductive delegate. For reward protocol, is it ok?

Reply:
For reward protocol, we may not need to call Delegates. You can try first.

Reply:
I'm trying right now, we do have candidate info from context, but not delegate info. before easter, we need to know unqualified delegate.

Reply:
I just tried to do that, but we need to Delegates() in reward protocol, because we need to calculate unproductive list before easter. they do have candidate from context but not delegate list.